### ISSUE TITLE
Screenshot portal needs to provide advanced features

### ISSUE THREAD

**Issue by Jehan** [flatpak/xdg-desktop-portal#139]
In GIMP, we are improving screenshot for Wayland and sandboxes, etc. We are being told that the screenshot portal is the way to go and that it should the portable API across all desktop environment and all sandboxes, which is very good (exactly what I was looking for).

As for features, I am being told that the API is not meant to expose any feature (cf. https://github.com/flatpak/flatpak/issues/1090#issuecomment-351068428). Advanced features will be presented to the user through the portal dialog and in the end a screenshot will simply be returned to GIMP.
So I made an implementation in GIMP with org.freedesktop.portal.Screenshot.

Unfortunately it seems the current implementation from xdg-desktop-portal-gtk package has no feature at all. It takes a screenshot, that's all. All the dialog proposes is if I accept to give the screenshot to the application. As a consequence, I set the screenshot portal as the last implementation, which will only be called if the GNOME API, KDE API and X11 backends are not found. In other words, it will likely never be used. Of course, I plan to make it the first implementation, but we need to have the basic features at least, otherwise that's just a huge regression.
The features are:

- Full screenshot (every displays as a single image, also called "root" or "full screen").
- Screenshot of a single display (for multi-display setup, you just want the content of a single screen).
- Screenshot of a single Window (ideally by active selection. GNOME has a "active window" version of this feature, but I find it a lot less usable than ability to click a window with a cursor, as does KDE API for instance).
- Screenshot of a rectangular selection.
- Include the window decorations: option available only for single window snaps.
- Include the pointer in the screenshot.
- Delay between window or rectangular selection and actual screenshot.
- Delay before screenshot for all displays and single display.

**Comment by Jehan** [comment]
Oh I completely forgot a very important feature! GIMP needs absolutely color management!
We need the screenshot feature to be color managed (a screen profile could be sent as an argument in the result vardict, but this would become complicated with multi-display setups; so probably the pixel colors have rather to  be converted from the their associated display profile to a known profile).

**Comment by matthiasclasen** [comment]
I'm not going to promise that the screenshot portal will provide every single feature you are asking for here. Some of these might be reasonable to add, if we can come up with a UX that does not hamper the 'just take a plain screenshot, no questions asked' workflow.

**Comment by Jehan** [comment]
> if we can come up with a UX that does not hamper the 'just take a plain screenshot, no questions asked' workflow

Maybe can't the API have just this one option: whether an app wants a "plain screenshot, no questions asked" or whether one want a featureful screenshot dialog?

**Comment by matthiasclasen** [comment]
yes, thats quite likely the api we would offer here: an "interactive" boolean option

**Comment by matthiasclasen** [comment]
https://github.com/flatpak/xdg-desktop-portal-gtk/pull/106

**Comment by jadahl** [comment]
How is the screenshot portal supposed to handle multiple monitors with different scaling? The only possible ways in the current API seems to either only take a screenshot of a single monitor, or to scale up every monitor to match the scale of the monitor with the highest scale, which involves choosing a scaling algorithm etc.

**Comment by matthiasclasen** [comment]
I've merged the interactive options now.

Wrt to multi-monitor, the screenshot portal is not providing more than existing solution like gnome-screenshot - ie no support

**Comment by Jehan** [comment]
Thanks!

Quickly skimming through the commit changes, there doesn't seem to be anything relative to color management as well. How can we get profile information from monitors in Wayland?

Not doing anything more than the current features, does it mean that you don't plan to do more and think that's all good as-is? Getting accurate color information is extremely important for graphics application (otherwise RGB information, that's just meaningless numbers basically!). Especially now that we have worked very hard to have color management in the core of GIMP and everywhere where colors are dealt with (i.e. nearly everywhere).

**Comment by matthiasclasen** [comment]
Do you get profile information via your various other screenshot fallbacks?

Not saying that there is no chance for this, but it is a bit of a niche feature, and you have to invest a lot of effort that will only pay of in very limited situations  (monitor is calibrated, the app you are screenshooting is paying careful attention to color correctness, you are not losing color precision on the way to or from the compositor...)

**Comment by Jehan** [comment]
> Do you get profile information via your various other screenshot fallbacks?

It depends.

-  The macOS implementation for screenshotting attachs a color profile directly with the screenshot result. So basically we have nothing to do and we have the right colors.
- Windows provides an API (GetICMProfile()) allowing to get the profiles of the various displays. That's still a bit of a problem when we get a snap of multi-displays so it's not perfect on multi-display setups.
- In X11, we get the _ICC_PROFILE_*  attached on each respective screen. This has the same issue as Windows regarding multi-display setups.

Basically macOS is the one doing the best job by just attaching a single profile to the resulting image. If that was a multi-display setup, I am assuming it just did the necessary conversion needed to represent all the pixels with a single color profile.

**Comment by Jehan** [comment]
> Not saying that there is no chance for this, but it is a bit of a niche feature

Also I would not say that is a niche. Well it's not the most basic person who calibrates one's screen, that's for sure. :-)
But designers, graphists, painters, but also video people, etc. I would not say that's a "small niche" at the very least. Basically that involves any person working with graphics and video. This is still quite a big market.

Now I'm not saying to make it a high priority. :-)
But just not discarding the feature as possibly addable in the future.
Which means also that if it needs to be part of the API, that should be now, even if just as a placeholder which will later be actually filled with the relevant profile for the screenshot.

Probably @hughsie could be part of this discussion.

**Comment by Jehan** [comment]
>  Which means also that if it needs to be part of the API, that should be now, even if just as a placeholder which will later be actually filled with the relevant profile for the screenshot.

Actually since the XDG API returns a file on disk, you can just embed the profile directly in the file. No need even to change anything in the API or add any option.
This should always be done, and then it is up to any application to do whatever it wants (not use the profile for display or even discard the profile, or actually manage colors and display the right colors).
